viewing_party/party.py

```diff
@@ -1,23 +1,129 @@
+    return sorted_genres[0][0] if len(sorted_genres) > 0 else None
 
 # -----------------------------------------
 # ------------- WAVE 3 --------------------
 # -----------------------------------------
+def get_unique_watched(user_data):
+    user_movies, friend_movies = set(), set()
+
+    for user_movie in user_data["watched"]:
+        user_movies.add(user_movie["title"])
+
+    for friend_dict in user_data["friends"]:
+        for friend_movie in friend_dict["watched"]:
+            friend_movies.add(friend_movie["title"])
+    unique_title = user_movies - friend_movies
+
+    unique_movies = []
+    for movie in user_data["watched"]:
+        if movie["title"] in unique_title and movie not in unique_movies:
+            unique_movies.append(movie)
+
+    return unique_movies
+
+def get_friends_unique_watched(user_data):
+
+    user_movies, friend_movies = set(), set()
+
+    for user_movie in user_data["watched"]:
+        user_movies.add(user_movie["title"])
+
+    for friend_dict in user_data["friends"]:
+        for friend_movie in friend_dict["watched"]:
+            friend_movies.add(friend_movie["title"])
+    friend_unique_titles = friend_movies - user_movies
+
+    friend_unique_movies = []
+    for friend_dict in user_data["friends"]:
+        for friend_movie in friend_dict["watched"]:    
+            if friend_movie["title"] in friend_unique_titles:
+                if friend_movie not in friend_unique_movies:
+                    friend_unique_movies.append(friend_movie)
+
+    return friend_unique_movies
 
-        
 # -----------------------------------------
 # ------------- WAVE 4 --------------------
 # -----------------------------------------
+def get_available_recs(user_data):
+
+    # if user_data["watched"] == []: return []
+    
+    streamings = set()
+    for service in user_data["subscriptions"]:
+        streamings.add(service)
+
+    recommended = []
+    for friend in user_data["friends"]:
+            for movie in friend["watched"]:
+                if movie in get_friends_unique_watched(user_data):
+                    if movie["host"] in streamings:
+                        recommended.append(movie)
+
+    return recommended
 
 # -----------------------------------------
 # ------------- WAVE 5 --------------------
 # -----------------------------------------
 
+def get_new_rec_by_genre(user_data):
+    most_freq_genre = get_most_watched_genre(user_data) 
+    rec_by_genre = []
+    for movie in get_friends_unique_watched(user_data):
+        if movie["genre"] == most_freq_genre:
+            rec_by_genre.append(movie)
+    return rec_by_genre
+
+def get_rec_from_favorites(user_data):
+    rec_movies = []
+    for fav_movie in user_data["favorites"]:
+        for uniq_movie in get_unique_watched(user_data):
+            if fav_movie["title"] == uniq_movie["title"]:
+                rec_movies.append(fav_movie)
+    return rec_movies

```

Comment:
We're nesting loops a little more than required here. We could remove the loop over `user_data["favorites"]` and keep the loop over the result of `get_unique_watched`. Then in each iteration, we'd check if the current move is in `user_data['favorites']`.
```python
for uniq_movie in get_unique_watched(user_data):
    if uniq_movie in user_data['favorites']:
        rec_movies.append(uniq_movie)

return rec_movies
```